### ISSUE TITLE
Add range 'for' with float step

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#4164]
Tested on Windows 7 64 bits with Godot 2.0.1:

The following code produces an error:

``` gdscript
    for t in range(-PI, PI, PI/4.0):
        print(str(t))
```

"step argument is zero!"
Because range() seems to only accept integers, so because PI/4.0 is below 1, it interprets it as 0, and produces the error.

It would be nice to have float support in range too :)

**Comment by bojidar-bg** [comment]
:+1:

**Comment by LeonardMeagher2** [comment]
Why not multiply by something

```gdscript
for t in range(-PI * 1000, PI * 1000, (PI/4)*1000):
   print(str(t/1000))
```

**Comment by OvermindDL1** [comment]
@LeonardMeagher2 Likely because you are both losing accuracy (not terribly a big deal) and you are converting float to int (not perfectly fast) at least once per loop as they likely want it as a float since it starts as floats.

It is a bit noisy, but a float range (or range of anything) can be done via a while loop:
```python
var t = -PI
while t <= PI:
  # Do stuff
  t += PI/4.0 # or doing PI*0.25 would technically run a little faster if it really matters, though the JIT should really do that itself...
```

**Comment by Zylann** [comment]
I use a while with a precalculated step at the moment. But a for would be easier to write and read.

**Comment by akien-mga** [comment]
It no longer produces an error in the master branch, but the step is still floored so the OP's example produces no output (as the step is 0).

```
    for t in range(-PI, PI, PI/2.0):
        print(str(t))
```
prints
```
-2
-1
0
1
2
```

As for the feature itself, I never needed it myself, I think it's better to handle such things yourself when you use floats, to avoid surprises with floating point precision. Yet, it should be relatively easy to implement if there's really demand for it.

**Comment by maciek134** [comment]
Wouldn't this be as simple as this:
```
func rangef(start: float, end: float, step: float):
  var res = Array()
  var i = start
  while i <= end:
    res.push_back(i)
    i += step
  return res
```
? Produces:
```
-3.141593
-2.356194
-1.570796
-0.785398
0
0.785398
1.570796
2.356194
3.141593
```

**Comment by bojidar-bg** [comment]
Does it work with this particularly contrived code?
```gdscript
var a = 0
var b = 0
for i in rangef(0.1, 0.3, 0.2):
  a += 1
for i in rangef(0.0, 0.2, 0.2):
  b += 1
assert(a == b) # Both should be 1 or both should be 2
```

**Comment by maciek134** [comment]
I didn't realize end was exclusive, fixed that to be in line with `range`. Now it works.

**Comment by aaronfranke** [comment]
Instead of this:

    for t in range(-PI, PI, PI/4.0):
        print(str(t))

It might be a good idea to have this:

    var pi_four = PI / 4 # For optimization purposes
    for t in range(-4, 4, 1):
        var x = t * pi_four
        print(str(x))

It probably runs faster (integer addition/comparison, then float multiplication, instead of float addition/comparison), avoids cumulative floating point precision errors, and makes it more clear how many steps there are above and below zero.

**Comment by maciek134** [comment]
Performance-wise they should be pretty similar - multiplication is pretty heavy and a loop on a float range doesn't have to do it. I don't have anything on the rounding error, other than it's pretty negligible for smaller numbers.

**Comment by akien-mga** [comment]
Feature and improvement proposals for the Godot Engine are now being discussed and reviewed in a dedicated [**Godot Improvement Proposals (GIP)**](https://github.com/godotengine/godot-proposals) ([godotengine/godot-proposals](https://github.com/godotengine/godot-proposals)) issue tracker. The GIP tracker has a detailed issue template designed so that proposals include all the relevant information to start a productive discussion and help the community assess the validity of the proposal for the engine.

The main ([godotengine/godot](https://github.com/godotengine/godot)) tracker is now solely dedicated to bug reports and Pull Requests, enabling contributors to have a better focus on bug fixing work. Therefore, we are now closing all older feature proposals on the main issue tracker.

If you are interested in this feature proposal, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). Be sure to reference this closed issue if it includes any relevant discussion (which you are also encouraged to summarize in the new proposal). Thanks in advance!

**Comment by aaronfranke** [comment]
Whoops, I forgot about this one. This was actually fixed by #37020. You can do this in master:

```gdscript
	for angle in Vector3(0.0, TAU, TAU / 20):
		print(angle)
```

Output:

```
0
0.314159
0.628319
0.942478
1.256637
1.570796
1.884956
2.199115
2.513274
2.827433
3.141593
3.455752
3.769911
4.084071
4.39823
4.712389
5.026548
5.340708
5.654867
5.96902
```

Removing "archived" label since this was fixed.
